lib/dot-compat.js

```diff
@@ -0,0 +1,298 @@
+
+    const dotConfig = {};
+    const configs = [];
+    const languageOptions = {};
+    const linterOptions = {};
+    const keysToCopy = ["settings", "rules", "processor"];
+    const languageOptionsKeysToCopy = ["globals", "parser", "parserOptions"];
+    const linterOptionsKeysToCopy = ["noInlineConfig", "reportUnusedDisableDirectives"];
+    const pluginEnvironments = new Map();
+
+    // check for special settings for eslint:all and eslint:recommended:
+    if (eslintrcConfig.settings) {
+        if (eslintrcConfig.settings["eslint:all"] === true) {
+            return ["eslint:all"];
+        }
+
+        if (eslintrcConfig.settings["eslint:recommended"] === true) {
+            return ["eslint:recommended"];
+        }
+    }
+
+    // copy over simple translations
+    for (const key of keysToCopy) {
+        if (key in eslintrcConfig && typeof eslintrcConfig[key] !== "undefined") {
+            dotConfig[key] = eslintrcConfig[key];
+        }
+    }
+
+    // copy over languageOptions
+    for (const key of languageOptionsKeysToCopy) {
+        if (key in eslintrcConfig && typeof eslintrcConfig[key] !== "undefined") {
+            dotConfig.languageOptions = languageOptions;
+
+            if (languageOptions[key] && typeof languageOptions[key] === "object") {
+                languageOptions[key] = {
+                    ...eslintrcConfig[key]
+                };
+            } else {
+                languageOptions[key] = eslintrcConfig[key];
+
+                if (key === "parser") {
+                    debug(`Resolving parser '${languageOptions[key]}' relative to ${resolveConfigRelativeTo}`);
+                    languageOptions[key] = eslintrcConfig[key].definition;
+                }
+            }
+        }
+    }
+
+    // copy over linterOptions
+    for (const key of linterOptionsKeysToCopy) {
+        if (key in eslintrcConfig && typeof eslintrcConfig[key] !== "undefined") {
+            dotConfig.linterOptions = linterOptions;
+            linterOptions[key] = eslintrcConfig[key];
+        }
+    }
+
+    // move ecmaVersion a level up
+    if (languageOptions.parserOptions) {
+
+        if ("ecmaVersion" in languageOptions.parserOptions) {
+            languageOptions.ecmaVersion = languageOptions.parserOptions.ecmaVersion;
+            delete languageOptions.parserOptions.ecmaVersion;
+        }
+
+        if ("sourceType" in languageOptions.parserOptions) {
+            languageOptions.sourceType = languageOptions.parserOptions.sourceType;
+            delete languageOptions.parserOptions.sourceType;
+        }
+
+        // check to see if we even need parserOptions anymore and remove it if not
+        if (Object.keys(languageOptions.parserOptions).length === 0) {
+            delete languageOptions.parserOptions;
+        }
+    }
+
+    if (eslintrcConfig.ignorePattern) {
+        configs.unshift({
+            ignores: eslintrcConfig.ignorePattern.patterns
+        });
```

Comment:
This will not work because the `ignorePatterns` value is not glob. The `ignorePatterns` value is `.gitignore`-style patterns. For examples of differences,

- negative patterns ... `!foo.js`
- slash at beginning ... `/foo.js` matches `foo.js` only on the top-level directory. On the other hand, `foo.js` matches `foo.js` on sub-directories as well.
- slash at ending ... `foo/` matches `foo` directory, but not file.
- dot ... `./foo.js` matches `foo.js` in the directory named `.`, not the current directory.

I'm not sure if other differences exist or not.

Reply:
It needs to care about negative patterns. In `.gitignore`-style patterns, we cannot unignore files in ignored directories. For example, `["foo/", "!bar.js"]` doesn't match `foo/bar.js`. `["foo/*", "!bar.js"]` *does* match `foo/bar.js`.

Reply:
Hmm okay, I'll need to think about this. It may be that we just can't correctly translate `ignorePatterns`.

Reply:
There are three reasons that I decided to use `.gitignore`-style patterns rather than glob patterns.

- Overridable. Because `.gitignore`-style patterns have negative patterns, we can override `ignorePatterns` properties of shareable configs by the user config's `ignorePatterns` property in spite of the property (string arrays) is simply concatenated.
- For smooth migration from `.eslintignore`. That's using `.gitignore`-style patterns. Therefore, people can migrate as (1) copy the patterns to `.eslintrc.*` and (2) remove the `.eslintignore` file.
- We have received many requests that people want to use the `.gitignore` file. In `.eslintrc.js`, we can use `fs.readFileSync(".gitignore", "utf8")` (or [eslint-gitignore](https://www.npmjs.com/package/eslint-gitignore) package) to use the `.gitignore` file without additional conversion.

Those pros are valid on the new flat config, I think.

----

How about the `ignores` property allows a function? Popular tools (e.g. babel, webpack, etc...) support functions to choose files generally.
That allows us to use any logic to decide ignored files. It will be useful for people who want to use `.gitignore`. Also, we can use it for this compatible layer.

```js
configs.unshift({
    // If ESLint should ignore the file, return `true`.
    ignores(filePath) {
        // Compute the final config for this file.
        // This filters config array elements by `files`/`excludedFiles` then merges the elements.
        // https://github.com/eslint/eslintrc/blob/b86d0c24b6fc3e941913469355ba77fccb8a47bc/lib/config-array/config-array.js#L475-L485
        const finalConfig = eslintrcConfigs.extractConfig(filePath)
        // Test the `ignorePattern` properties of the final config.
        return finalConfig.ignores(filePath)
    }
})
```

Reply:
Yeah, I thought about allowing functions in `ignores`. The only bad thing about that is that, with strings, you can pull out all of the ignore patterns and apply them using something like `globby`. If we start putting functions in there, we lose that capability.

I need to look a bit closer at this part.

Reply:
Okay, it seems like allowing functions in the `ignores` array is the least bad solution to this problem. I'll get to work on that.